contracts/distribution/dao-rewards-distributor/src/contract.rs

```diff
@@ -0,0 +1,683 @@
+            amount: vec![Coin { denom, amount }],
+        }
+        .into()),
+        Denom::Cw20(addr) => {
+            let cw20_msg = to_json_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                recipient: recipient.into_string(),
+                amount,
+            })?;
+            Ok(WasmMsg::Execute {
+                contract_addr: addr.into_string(),
+                msg: cw20_msg,
+                funds: vec![],
+            }
+            .into())
+        }
+    }
+}
+
+pub fn update_rewards(
+    deps: &mut DepsMut,
+    env: &Env,
+    addr: &Addr,
+    denoms: Vec<String>,
+) -> StdResult<()> {
+    for denom in denoms {
+        let reward_config = REWARD_DENOM_CONFIGS.load(deps.storage, denom.clone())?;
+
+        // first, we calculate the rewards per token and update them
+        let rewards_per_token = get_rewards_per_token(
+            &reward_config,
+            env,
+            &reward_config.vp_contract,
+            deps.as_ref(),
+        )?;
+
+        // update the cumulative rewards per token with latest rewards per token
+        CUMULATIVE_REWARDS_PER_TOKEN.save(deps.storage, denom.clone(), &rewards_per_token)?;
+
+        // then we calculate the rewards earned since last user action
+        let earned_rewards = get_accrued_rewards_since_last_user_action(
+            deps.as_ref(),
+            env,
+            addr,
+            rewards_per_token,
+            &reward_config.vp_contract,
+            vec![denom.clone()],
+        )?;
+
+        // reflect the earned rewards in the user's reward config
+        USER_REWARD_CONFIGS.update(deps.storage, addr.clone(), |config| -> StdResult<_> {
+            // if user does not yet have a config, we create a new one
+            let mut user_reward_config = config.unwrap_or_default();
+
+            // get the pre-existing pending reward amount for the denom
+            let previous_pending_denom_reward_amount = *user_reward_config
+                .pending_denom_rewards
+                .get(&denom)
+                .unwrap_or(&Uint128::zero());
+
+            // get the amount of newly earned rewards for the denom
+            let earned_rewards_amount = earned_rewards.get(&denom).cloned().unwrap_or_default();
+
+            user_reward_config.pending_denom_rewards.insert(
+                denom.clone(),
+                previous_pending_denom_reward_amount + earned_rewards_amount,
+            );
+
+            user_reward_config
+                .user_reward_per_token
+                .insert(denom.clone(), rewards_per_token);
+
+            Ok(user_reward_config)
+        })?;
+
+        // Update the last update expiration in the DenomRewardConfig
+        REWARD_DENOM_CONFIGS.update(deps.storage, denom.clone(), |config| -> StdResult<_> {
+            match config {
+                Some(rc) => Ok(rc.bump_last_update(&env.block)),
+                None => Err(StdError::generic_err("Denom config not found")),
+            }
+        })?;
```

Comment:
merged the CUMULATIVE_REWARDS_PER_TOKEN value into the reward denom map since they were always tied together. it also makes it more clear that the `last_update` field in the denom reward struct is talking about the `total_earned_puvp` value